types/src/v0/impls/l1.rs

```diff
@@ -64,180 +75,470 @@ impl L1BlockInfo {
+                };
+                if finalized.timestamp < timestamp {
+                    tracing::debug!(%timestamp, ?finalized, "waiting for L1 finalized block");
+                    continue;
+                }
+                break 'outer (self.state.lock().await, finalized);
+            }
+
+            // This should not happen: the event stream ended. All we can do is try again.
+            tracing::warn!(%timestamp, "L1 event stream ended unexpectedly; retry",);
+            sleep(self.retry_delay).await;
+        };
+
+        // Find the earliest finalized block in our cache with timestamp greater or equal to
+        // `timestamp`.
+        let block = match state.finalized_by_timestamp.range(timestamp..).next() {
+            Some((_, number)) => {
+                let number = *number;
+                state.finalized.get(&number).copied()
             }
-            tracing::info!(
-                %timestamp,
-                ?block,
-                "waiting for finalized block with sufficient timestamp"
-            );
-            sleep(interval).await;
+            None => None,
         };
+        // There should almost always be something in the cache, but just in case (e.g. block
+        // getting evicted at just the wrong time), we can fall back to using the finalized
+        // snapshot, since we have previously verified that it is at least `timestamp`.
+        let mut block = block.unwrap_or(finalized);
 
-        // The finalized block jumps by more than 1 at a time, so we might not have found the
-        // earliest block with the desired timestamp. Work backwards until we find it.
+        // It is possible there is some earlier block that also has the proper timestamp; maybe we
+        // just didn't have it in cache. Work backwards until we find the true earliest block.
         loop {
-            let parent = self.wait_for_block(block.number - 1).await;
+            let (state_lock, parent) = self.get_finalized_block(state, block.number - 1).await;
             if parent.timestamp < timestamp {
                 return block;
             }
+            state = state_lock;
             block = parent;
         }
     }
 
-    /// Proxy to `Provider.get_block_number`.
-    async fn get_block_number(&self) -> u64 {
-        loop {
-            match self.provider.get_block_number().await {
-                Ok(n) => return n.as_u64(),
-                Err(e) => {
-                    tracing::warn!("Blocknumber error: {}", e);
+    async fn get_finalized_block<'a>(
+        &'a self,
+        mut state: MutexGuard<'a, L1State>,
+        number: u64,
+    ) -> (MutexGuard<'a, L1State>, L1BlockInfo) {
+        // Try to get the block from the finalized block cache.
+        assert!(
+            state.snapshot.finalized.is_some()
+                && number <= state.snapshot.finalized.unwrap().number,
+            "requesting a finalized block {number} that isn't finalized; snapshot: {:?}",
```

Comment:
Wonder why we don't make this a fallible function instead of one that might panic.

Reply:
The idea is this can't panic based solely on the internal behavior of this module, i.e. we only ever call this private helper function with a finalized block number. There is no input from callers in any other modules that can cause this function to panic, so we shouldn't bubble the error up to the caller.

And because the eventual interface is infallible (e.g. `validate_apply_header` calls `wait_for_finalized_block` and blocks until we either get it or time out) any errors we return would have to be handled with retry loops which IMO makes the calling code more complicated for little reason, when it's an error that should never happen to begin with